server/cloudapps/install_app.go

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package cloudapps
+
+import (
+	"errors"
+
+	"github.com/mattermost/mattermost-plugin-cloudapps/server/utils/md"
+)
+
+type InInstallApp struct {
+	ActingMattermostUserID string
+	NoUserConsentForOAuth2 bool
+	Manifest               *Manifest
+	Secret                 string
+}
+
+type OutInstallApp struct {
+	md.MD
+	App *App
+}
+
+func (r *registry) InstallApp(in *InInstallApp) (*OutInstallApp, error) {
+	if in.Manifest.AppID == "" {
+		return nil, errors.New("app ID must not be empty")
+	}
+	// TODO check if acting user is a sysadmin
+
+	// TODO remove mock, implement for real
```

Comment:
Regarding real implementation (to make sure we are all aligned):

We already got the app secret during the dialog, so we can send messages to the app, but we still need to inform the app of two things:
  - We are installed
  - What is our secret so they can communicate with us (or the OAuth token, or whichever communication we want to set in place)

I understand this will be done by a simple HTTP call to the app, or opening a new window to trigger the OAuth flow.

Reply:
Yes.

Reply:
> What is our secret so they can communicate with us (or the OAuth token, or whichever communication we want to set in place)

To use the REST API as a user, the App must go through the OAuth2 loop. To use the Bot-level token - TBD, but I think we discussed the general outline - provision the bot upon install, then use the bot account.

Reply:
In theory, the OAuth protocol should handle re-validation, but I have to dig deeper into that.